File: rate-limiter/src/limiter/tokenBucket.ts
import redis from "../config/redis";
import { RateLimiter } from "./rateLimiter";
import { RateLimitPolicy } from "../types/policy";
import { RateLimitResult } from "../types/decision";

export class TokenBucketLimiter implements RateLimiter {
  async consume(
    key: string,
    policy: RateLimitPolicy
  ): Promise<RateLimitResult> {
    const now = Math.floor(Date.now() / 1000);
    const bucketKey = `token_bucket:${key}`;

    const refillRate = policy.limit / policy.windowSeconds;

    const data = await redis.hgetall(bucketKey);

    let tokens = data.tokens ? parseFloat(data.tokens) : policy.limit;
    let lastRefillTs = data.lastRefillTs
      ? parseInt(data.lastRefillTs, 10)
      : now;

    // Refill tokens based on elapsed time
    const elapsed = now - lastRefillTs;
    const refill = elapsed * refillRate;

    tokens = Math.min(policy.limit, tokens + refill);

    let allowed = false;

    if (tokens >= 1) {
      tokens -= 1;
      allowed = true;
    }

    await redis.hset(bucketKey, {
      tokens: tokens.toString(),
      lastRefillTs: now.toString(),
    });

    // Optional cleanup
    await redis.expire(bucketKey, policy.windowSeconds * 2);

    return {
      allowed,
      remaining: Math.floor(tokens),
      resetAt: now + Math.ceil((policy.limit - tokens) / refillRate),
    };
  }
}

File: rate-limiter/src/limiter/tokenBucket.lua
-- KEYS[1] = bucket key
-- ARGV[1] = capacity
-- ARGV[2] = refill rate (tokens per second)
-- ARGV[3] = current timestamp (seconds)
-- ARGV[4] = ttl (seconds)

local bucket = redis.call("HGETALL", KEYS[1])

local tokens = tonumber(ARGV[1])
local lastRefill = tonumber(ARGV[3])

if #bucket > 0 then
  for i = 1, #bucket, 2 do
    if bucket[i] == "tokens" then
      tokens = tonumber(bucket[i + 1])
    elseif bucket[i] == "lastRefillTs" then
      lastRefill = tonumber(bucket[i + 1])
    end
  end
end

local now = tonumber(ARGV[3])
local refillRate = tonumber(ARGV[2])

local elapsed = now - lastRefill
local refill = elapsed * refillRate
tokens = math.min(tonumber(ARGV[1]), tokens + refill)

local allowed = 0
if tokens >= 1 then
  tokens = tokens - 1
  allowed = 1
end

redis.call("HSET", KEYS[1],
  "tokens", tokens,
  "lastRefillTs", now
)

redis.call("EXPIRE", KEYS[1], ARGV[4])

return { allowed, tokens }
